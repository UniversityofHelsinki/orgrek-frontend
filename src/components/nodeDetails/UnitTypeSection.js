import React from 'react';
import EditableAccordion from '../EditableAccordion';
import { useTranslation } from 'react-i18next';
import AttributesTable from '../attributes/AttributesTable';
import Placeholder from '../Placeholder';
import EditableContent from '../EditableContent';
import UnitTypeEditor from './UnitTypeEditor';
import {
  useGetTypeAttributesQuery,
  useGetValidHierarchyFiltersQuery,
  useSaveTypeAttributesMutation,
} from '../../store';
import { useNodeId } from '../../hooks/useNodeId';
import useSortAttributesByDate from '../../hooks/useSortAttributesByDate';
import { useSelector } from 'react-redux';
import fillSelectableUnits from '../../hooks/filterSelectableUnits';
import { authActions } from '../../auth';
import { compareAndCheckDates, valueNotEmpty } from './Validations';

const UnitTypeSection = () => {
  const { t } = useTranslation();
  const nodeId = useNodeId();
  const { data, error, isFetching } = useGetTypeAttributesQuery(nodeId);
  const {
    data: hierarchies,
    hierarchyerror,
    isFetchingHierarchy,
  } = useGetValidHierarchyFiltersQuery();
  const selectedHierarchies = useSelector(
    (state) => state.tree.selectedHierarchy
  );
  const [saveTypeAttributes] = useSaveTypeAttributesMutation();
  const selectableUnits = [];

  // In edit mode data includes also history and future
  const sortedData = useSortAttributesByDate(data);

  const title = t('unit_type');
  const empty = sortedData.length === 0;

  const toFormValues = (data) => {
    const foundTypes = [];
    fillSelectableUnits(selectableUnits, hierarchies, selectedHierarchies);
    data.forEach((value) => {
      selectableUnits.forEach((o) => {
        if (o.value === value.value) {
          foundTypes.push(value);
        }
      });
    });
    return { type: [...foundTypes] };
  };

  const { type } = toFormValues(sortedData);

  // Validates form values every time when the values change
  // Submit button is disabled when errors contain any truthy values
  // EditableContent handles displaying form-level validation error messages
  const validate = (values) => {
    return {
      ...valueNotEmpty(values.type),
      ...compareAndCheckDates(values.type),
    };
  };

  const ObjetToArray = (obj) => Object.assign([], Object.values(obj));

  const handleSubmit = (values) => {
    const valuesArray = ObjetToArray(values.type);
    return saveTypeAttributes({ valuesArray, nodeId }).unwrap();
  };

  return (
    <EditableAccordion
      title={title}
      loading={isFetching}
      defaultExpanded={!empty}
    >
      <EditableContent
        editorComponent={<UnitTypeEditor />}
        validate={validate}
        initialValues={toFormValues(sortedData)}
        onSubmit={handleSubmit}
        successMessage={t('typeInfo.saveSuccess')}
        errorMessage={t('typeInfo.saveError')}
        authActions={authActions.unitType}
      >
        <Placeholder empty={empty} placeholder={t('typeInfo.empty')}>
          <AttributesTable data={[...type]} summary={title} />
        </Placeholder>
      </EditableContent>
    </EditableAccordion>
  );
};

export default UnitTypeSection;
