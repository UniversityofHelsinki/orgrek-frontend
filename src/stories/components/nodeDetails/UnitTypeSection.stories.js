import {
  hierarchyFilters,
  mockGetAttributes,
  mockGetSectionTypeAttributes,
  mockGetTypeAttributes,
  mockGetValidHierarchyFilters,
  mockPutTypeAttributes,
  withMockStore,
} from '../../../mockStore';
import UnitTypeSection from '../../../components/nodeDetails/UnitTypeSection';
import { waitFor, within } from '@storybook/test';
import { expect } from '@storybook/test';

// Use a fixed date to ensure that tests always have a consistent result
const now = new Date('2023-03-22T14:28:00+0200');

const nodeId = '1';

export default {
  component: UnitTypeSection,
  parameters: {
    systemTime: now,
    reactRouter: {
      searchParams: {
        uid: nodeId,
        hierarchies: 'virallinen',
      },
    },
  },
};

const sectionTypeData = [
  {
    id: 17,
    section: 'types',
    attr: 'type',
  },
];

export const Default = {
  parameters: {
    systemTime: now,
    msw: {
      handlers: [
        mockGetAttributes(nodeId, {
          types: [
            {
              id: 1,
              nodeId: '1',
              key: 'type',
              value: 'tiedekunta',
              startDate: null,
              endDate: null,
            },
          ],
        }),
        mockPutTypeAttributes(nodeId),
        mockGetValidHierarchyFilters(now, hierarchyFilters),
        mockGetSectionTypeAttributes('types', sectionTypeData),
      ],
    },
  },
  decorators: [withMockStore({ tree: { selectedHierarchy: 'virallinen' } })],
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('tiedekunta')).toBeInTheDocument();
    });
  },
};

export const Empty = {
  ...Default,
  parameters: {
    msw: {
      handlers: [
        mockGetAttributes(nodeId, { types: [] }),
        mockPutTypeAttributes(nodeId),
        mockGetValidHierarchyFilters(now, hierarchyFilters),
        mockGetSectionTypeAttributes('types', sectionTypeData),
      ],
    },
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    await waitFor(async () => {
      await expect(canvas.getByText('Yksik√∂n tyyppi')).toBeInTheDocument();
    });

    await expect(canvas.getByRole('button')).toHaveAttribute(
      'aria-expanded',
      'false'
    );
  },
};
