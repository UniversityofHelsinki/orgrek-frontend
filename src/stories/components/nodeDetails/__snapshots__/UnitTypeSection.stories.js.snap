// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/nodeDetails/UnitTypeSection Default smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "expanded": true,
      "name": "unit_type",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "edit_mode_edit_button",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "attribute",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "value",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "valid_dates",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "type",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "tiedekunta",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "not_specified",
      "pressed": undefined,
      "role": "text",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-unittypesection--default",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;

exports[`stories/components/nodeDetails/UnitTypeSection Empty smoke-test 1`] = `
Object {
  "checked": undefined,
  "children": Array [
    Object {
      "checked": undefined,
      "children": undefined,
      "name": "unit_type",
      "pressed": undefined,
      "role": "button",
      "value": undefined,
    },
  ],
  "name": "stories-components-nodedetails-unittypesection--empty",
  "pressed": undefined,
  "role": "WebArea",
  "value": undefined,
}
`;
